### PR TITLE
Add sonar gradle plugin, append jacoco to single file

### LINKED ISSUE #190: Run sonar analysis at continuous integration
https://sonarqube.com/projects is free for open-source project analysis.
We should run sonar analyzer at CI build to see the project code measures like code coverage and code complexity etc

### PR DESCRIPTION
**Pull request by kzaikin**
Add sonarqube plugins, set it up to work with sonarcloud.io
Project is analyzed at full builds only, at pull requests sonar is disabled.
To give sonar unit test coverage data, coverage for all projects is appended to single file.
Coveralls accepts this kind of coverage just fine too.

fix #190

### CODE REVIEW

File: gradle/sonar.gradle [hunk 1]
```diff
@@ -0,0 +1,9 @@
+apply plugin: 'org.sonarqube'
+sonarqube {
+    properties {
+        property 'sonar.host.url', System.env.SONAR_HOST_URL
+        property 'sonar.projectKey', 'edu.umd.cs.findbugs'
```
Comment (kzaikin) [inline gradle/sonar.gradle]: or should we use something like ''com.github.spotbugs'?

File: spotbugsTestCases/build.gradle [hunk 2]
```diff
@@ -41,3 +41,7 @@ task classesJava7(type:JavaCompile) {
 }
 
 tasks['build'].dependsOn classesJava7
+
+sonarqube {
+  skipProject true
```
Comment (kzaikin) [inline spotbugsTestCases/build.gradle]: this project should not be in sonarqube as it is test data, not real code

File: .travis.yml [hunk 3]
```diff
@@ -4,6 +4,8 @@ jdk: oraclejdk8
 
 env:
   global:
+    - SONAR_HOST_URL="https://sonarcloud.io"
+#TODO    - SONAR_LOGIN=
```
Comment (kzaikin) [inline .travis.yml]: @KengoTODA you have already added SpotBugs project to sonarcloud.io, so could you please give me a sonar login token as a secret prepared with Travis?

Comment (KengoTODA) [inline .travis.yml]: Please have a try with `secure: "N/wdh1L9n2kp+BfM8RfmOQQIlbngTKlO5EmnlMh7Q8RXK3N7U/k+74r7yRCqWNJd4zpzKpyQYKYxX6ZkwH8bcJ/SOipDgC9p4yrjY+f+juDK0LrWHg1pqssJL5Q7NTMlaXl5E8q54mCn59t//JrNC2+/KrjoYbJLaRyhaEe0nolAJsxhP1oEjs19Y+QRk9jbpByQLesbVy5LKekxaXQ94wfuPSJPcLDBYh9yQ9aMorqW9Pn2ZZpm7e0AlYjstUoShZY8nb8gsNNpvxrVme9q/1k3POxjbhnauwolc9WNvsFl8UdutCyZsmQbDNEZyJ/R9c9rtrJ5tT/bCZ34zpUX3wH9MjAb2wt+DFHZ7hZT8O5O5lQ1GG6fUZZBm4mT0JNg9HLPRk/TlVDJlmNm4KgzayTTqkD+ugVHNg7CBES5s8Ww9EM+7P2T74/8ZH0UwSBiEWM/cc3vYEpEThnPD5HcUMFGRFJJFAf1UCQAypc5mWFTlCIXuno34a8KHwAS5fyxhHGrIcLQZd1Sq+RIWnKQNekALZMRHQcDB53PGBk4GOtWmf4jAyWBxby/O52g4sJVtoX5jeVnVVPXwSXF2hJFnWDlvLAbV8nNU/PvqjXxS3h11/o11CPwu/vOeHzM5OWTqyCYttKQramr6/qALfW9DIiCr5D+mNu1CETvWFO0FGQ="`

File: spotbugsTestCases/build.gradle (hunk 2 above)
Comment (KengoTODA) [inline spotbugsTestCases/build.gradle]: let's comment in this file, to tell your intent to developers even in the future

File: gradle/jacoco.gradle [hunk 4]
```diff
@@ -10,3 +10,11 @@ jacocoTestReport {
     html.enabled = true
   }
 }
+
+test {
+  jacoco {
```
Comment (kzaikin) [inline gradle/jacoco.gradle]: this let us collect all coverage data to single file during build

File: .travis.yml [hunk 5]
```diff
@@ -4,6 +4,11 @@ jdk: oraclejdk8
 
 env:
   global:
+    - SONAR_HOST_URL="https://sonarcloud.io"
+
+    # SONAR_LOGIN value
+    - secure: "N/wdh1L9n2kp+BfM8RfmOQQIlbngTKlO5EmnlMh7Q8RXK3N7U/k+74r7yRCqWNJd4zpzKpyQYKYxX6ZkwH8bcJ/SOipDgC9p4yrjY+f+juDK0LrWHg1pqssJL5Q7NTMlaXl5E8q54mCn59t//JrNC2+/KrjoYbJLaRyhaEe0nolAJsxhP1oEjs19Y+QRk9jbpByQLesbVy5LKekxaXQ94wfuPSJPcLDBYh9yQ9aMorqW9Pn2ZZpm7e0AlYjstUoShZY8nb8gsNNpvxrVme9q/1k3POxjbhnauwolc9WNvsFl8UdutCyZsmQbDNEZyJ/R9c9rtrJ5tT/bCZ34zpUX3wH9MjAb2wt+DFHZ7hZT8O5O5lQ1GG6fUZZBm4mT0JNg9HLPRk/TlVDJlmNm4KgzayTTqkD+ugVHNg7CBES5s8Ww9EM+7P2T74/8ZH0UwSBiEWM/cc3vYEpEThnPD5HcUMFGRFJJFAf1UCQAypc5mWFTlCIXuno34a8KHwAS5fyxhHGrIcLQZd1Sq+RIWnKQNekALZMRHQcDB53PGBk4GOtWmf4jAyWBxby/O52g4sJVtoX5jeVnVVPXwSXF2hJFnWDlvLAbV8nNU/PvqjXxS3h11/o11CPwu/vOeHzM5OWTqyCYttKQramr6/qALfW9DIiCr5D+mNu1CETvWFO0FGQ="
```
Comment (kzaikin) [inline .travis.yml]: @KengoTODA I still get 'You're only authorized to execute a local (preview) SonarQube analysis without pushing the results to the SonarQube server' with this line added.

Did you hide SONAR_LOGIN=<token> in there?
I look for `System.env.SONAR_LOGIN` from gradle

Comment (kzaikin) [inline .travis.yml]: I've added the check and I can see that SONAR_LOGIN environment variable is not available during travis build

Comment (kzaikin) [inline .travis.yml]: @KengoTODA who have access to travis-ci and who would have access to project in sonar?

Comment (KengoTODA) [inline .travis.yml]: > Did you hide SONAR_LOGIN=token_value in there?
Maybe not, I'll regenerate it

Comment (KengoTODA) [inline .travis.yml]: @kzaikin Could you have a try with following parameter? I've hidden `SONAR_LOGIN=token` in this text:

>"FjzPYgjdN9WRvAwvJDFP0MqWMOsqCG0tNO6ts7X1p6HWikpoHnTqWgMl0YOgGK93trvB6vtIisZVVdLZG1rCwnXlO4Xd+6f+nhsJDledd9b4EePQPECJbHZZ8Hg69ATIlm7chll73GZzWTAL2Q3pIdsEJbwtqPzJ/j0H4COkayDeQ4ERA50pvvkXTiQrwxc5URrNH36ZOn10hfUEKO03EKbAlD+5dOzcxqu5WiS7j10zbMBCrKqhzN+zJQSOuDCTQz2S8McJBTiz4kMEBkF1q4Ax+1ywz2vICL3/F5/xqnHvHOk9wZE/4OOiNxjqxVd+Ik6LGE1BbLuATQz0Cmc3DiqsGYBsKtlk88v1QX2hsND/K0hwda/uItzbh95HmnrRYHYhzOcgcIVRqVwoPkYYWPYr9S2f3DGlFEnlGwJHgvqjuwMJaBmm2zblxPammEx3XoWqrKQ7RrelH1tUTWXy/tRss4Fof+D33/NJqa71Xvph1meKMcu+OOVatCul1IJaTtf2pyyifLTNRntgl4C4ok57PN882lRp55/Ifmi+xbt3Nd7lM1Fpy2eDugK1UINmkjRQsyCZY3EVgcM/NkrMa6qHhZFu5F1iv3EBYRiV/WKJ9PtTErkTeX1JZtyZYo7yS+VlIXIJFLoxfrV9TH5B3kczVILxgfJJ6gVPQOY5984="

File: gradle/sonar.gradle [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+//TODO(kzaikin): use preview mode in pull requests
+//if (new Boolean(System.env.TRAVIS_PULL_REQUEST)) {
+//    analysisMode = 'preview'
+//}
+
+sonarqube {
+    properties {
+        property 'sonar.analysis.mode', analysisMode
+        property 'sonar.host.url', System.env.SONAR_HOST_URL
+        property 'sonar.login', '250cb5c72d59d12620f1733fc7f397850d76ade0'
```
Comment (kzaikin) [inline gradle/sonar.gradle]: it's a temporary value just to check online analysis

Comment (kzaikin) [inline gradle/sonar.gradle]: @KengoTODA please take a look at temporary analyzed project
https://sonarcloud.io/dashboard?id=kzaikin.com.github.spotbugs.spotbugs

Looks like we must define sonar organization to analyze project. It's a new feature for sonarcloud.
Did you create a Spotbugs organization at sonarqube? Could you tell me its name for I to specify it here?

Comment (KengoTODA) [inline gradle/sonar.gradle]: I've added you to member of spotbugs organization, please check [member page](https://sonarcloud.io/organizations/spotbugs/members).

File: gradle/sonar.gradle [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+//TODO(kzaikin): use preview mode in pull requests
+//if (new Boolean(System.env.TRAVIS_PULL_REQUEST)) {
+//    analysisMode = 'preview'
+//}
+
+sonarqube {
+    properties {
+        property 'sonar.analysis.mode', analysisMode
+        property 'sonar.host.url', System.env.SONAR_HOST_URL
+        property 'sonar.login', '11cab05437bb5e07221bd50afc1e694ea8f64b49'
```
Comment (kzaikin) [inline gradle/sonar.gradle]: as soon as I have specified this temporary sonar token I was able to run sonar analysis of the project:
https://sonarcloud.io/dashboard?id=com.github.spotbugs.spotbugs

Looks like we have some problems with travis-ci secret values. What about specifying SONAR_LOGIN environment variable in travis-ci's project properties?

Comment (KengoTODA) [inline gradle/sonar.gradle]: Maybe possible at https://travis-ci.org/spotbugs/spotbugs/settings

But it's not so good habit, I will also try to use travis-ci secret value.

Comment (KengoTODA) [inline gradle/sonar.gradle]: Anyway, I've added `SONAR_LOGIN` environment variable, please have a try. @kzaikin

File: .travis.yml [hunk 8]
```diff
@@ -4,6 +4,13 @@ jdk: oraclejdk8
 
 env:
   global:
+    - SONAR_HOST_URL="https://sonarcloud.io"
+
+    # SONAR_LOGIN value
+    # please note this value is available for pull requests from the same repository only
+    # see https://docs.travis-ci.com/user/pull-requests/#Pull-Requests-and-Security-Restrictions
```
Comment (kzaikin) [inline .travis.yml]: Ok I got it.
My pull request is from fork to main repository
That is why travis-ci is hiding secret values
In fact sonar is useless at pull requests until there is github commenting plugin is available at sonarcloud.io.
See my next comment

File: gradle/sonar.gradle [hunk 9]
```diff
@@ -0,0 +1,18 @@
+apply plugin: 'org.sonarqube'
+
+// disable sonar for pull request builds until
+// sonar plugin for pull request comments is available at sonarcloud.io
+if (!new Boolean(System.env.TRAVIS_PULL_REQUEST)) {
+    rootProject.tasks['sonarqube'].setEnabled(false)
```
Comment (kzaikin) [inline gradle/sonar.gradle]: I disable sonar in pull requests intentionally
Default analysis mode analyses whole project and uploads analysis into sonar pretenting it to be the current state of the project.
So pull requests are usually anlysed in 'preview' mode along with a special sonar plugin commenting found issues in pull request. Alas, sonarcloud.io does not have such a plugin so pull requests analysis is useless. That is why I disable it

File: .travis.yml (hunk 8 above)
Comment (kzaikin) [inline .travis.yml]: @KengoTODA you can remove SONAR_LOGIN variable from travis-ce project settings. This secret should be enough.
WARNING we have to merge this PR to see how full build with sonar analysis is performed.
I will be online throughout the day. You can find me in telegram as kzaikin for quick fixes should it be needed

Comment (KengoTODA) [inline .travis.yml]: > you can remove SONAR_LOGIN variable from travis-ce project settings. 

Thanks, I've removed that.

File: gradle/sonar.gradle [hunk 10]
```diff
@@ -0,0 +1,18 @@
+apply plugin: 'org.sonarqube'
+
+// disable sonar for pull request builds until
+// sonar plugin for pull request comments is available at sonarcloud.io
+if (!new Boolean(System.env.TRAVIS_PULL_REQUEST)) {
```
Comment (KengoTODA) [inline gradle/sonar.gradle]: According to [javadoc](https://docs.oracle.com/javase/8/docs/api/java/lang/Boolean.html#Boolean-java.lang.String-), `Boolean::new` will return true if and only if given string is `"true"`. But `TRAVIS_PULL_REQUEST` is always `"false"` or `PR number`, so this `!new Boolean(System.env.TRAVIS_PULL_REQUEST)` is always true.

And I think it's better to use `TRAVIS_BRANCH` instead, then we can avoid analysis even for the branch which is pushed to `spotbugs` project directly.

**Comment by KengoTODA** [comment]
By [this travis build](https://travis-ci.org/spotbugs/spotbugs/builds/247862792), I've confirmed that secure value is working as expected (I've pushed this branch to spotbugs repo, and disable `if` in `sonar.gradle`).

**Comment by kzaikin** [comment]
This PR is ready to merge

**Pull request merged**